sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -498,9 +592,26 @@ def fit(self, X, y, sample_weight=None):
                 )
 
             # Update gradients and hessians, inplace
-            self._loss.update_gradients_and_hessians(
-                gradients, hessians, y_train, raw_predictions, sample_weight_train
-            )
+            # Note that self._loss expects shape (n_samples,) for
+            # n_trees_per_iteration = 1 else shape (n_samples, n_trees_per_iteration).
+            # T (transpose) returns a view.
+            if self._loss.constant_hessian:
+                self._loss.gradient(
+                    y_true=y_train,
+                    raw_prediction=raw_predictions.T,
```

Comment:
The docstring for `_loss.gradient` says is needs `C` order arrays, but this array is F ordered.

```python
import numpy as np

x = np.zeros((10, 30), dtype=np.float32)
x.T.flags["F_CONTIGUOUS"]
# True
```

In the implementation accepts both memory layouts:

https://github.com/scikit-learn/scikit-learn/blob/3e0f49b62339941722d0dfbcb90f5af0ad1b2e6c/sklearn/_loss/_loss.pyx.tp#L1065

Are there performance concerns with using F-array in the `self._loss.gradient` call?

Reply:
The docstring of `BaseLoss` says
```python
"""
raw_prediction : C-contiguous array of shape (n_samples,) or array of shape (n_samples, n_classes)
"""
```
The meaning is:
- If shape `(n_samples,)`, then the array must be contiguous (C and F are equivalent in this case)
- If shape `(n_samples, n_classes)`, then no assumptions are made about contiguity.

I tested different settings and were not able to identify clear performance benefits with contiguous arrays in the multiclass case.

Reply:
Interesting, I thought there would have been some cpu cache performance issue with either "C" or "F" order. (Depending on which axis prange is iterating over)

Reply:
The `prange` loop is always over samples. For all losses except multinomial loss, samples are required to be contiguous (1d arrays, therefore C=F contiguous).
For the multinomial loss, the present HGBT required arrays to be contiguous in samples. This PR relaxes this requirement, as I found no clear performance gain (but this PR saves a loop over classes and is therefore faster).